lib/dashing/app.rb

```diff
@@ -70,13 +70,23 @@ def authenticated?(token)
   redirect "/" + dashboard
 end
 
+
 get '/events', provides: 'text/event-stream' do
   protected!
   response.headers['X-Accel-Buffering'] = 'no' # Disable buffering for nginx
-  stream :keep_open do |out|
-    settings.connections << out
+  stream do |out|
     out << latest_events
-    out.callback { settings.connections.delete(out) }
+    settings.connections << connection = {out: out, mutex: Mutex.new, terminated: false}
+    terminated = false
+
+    loop do
+      connection[:mutex].synchronize do
+        terminated = true if connection[:terminated]
+      end
+      break if terminated
+    end
+
```

Comment:
The idea that Sinatra doesn't implement rack hijack properly comes from smart dudes comments here: https://github.com/sinatra/sinatra/issues/1035#issuecomment-203485499

But a some of it is over my head.